vpr/src/route/connection_box_lookahead_map.cpp

```diff
@@ -30,361 +35,315 @@
-        const std::vector<int>& channel_node_list = get_rr_node_indcies(
-            rr_type_, start_x_, start_y_);
-
-        for (; index_ < channel_node_list.size(); index_++) {
-            int node_ind = channel_node_list[index_];
-
-            if (node_ind == OPEN || device_ctx.rr_nodes[node_ind].capacity() == 0) {
-                continue;
-            }
-
-            const std::pair<size_t, size_t>* loc = device_ctx.connection_boxes.find_canonical_loc(node_ind);
-            if (loc == nullptr) {
-                continue;
+    for (size_t box_id = 0;
+         box_id < device_ctx.connection_boxes.num_connection_box_types();
+         box_id++) {
+        auto& matrix = cost_map_[iseg][box_id];
+        if (matrix.dim_size(0) == 0 || matrix.dim_size(1) == 0) {
+            printf("cost EMPTY for box_id = %lu\n", box_id);
+            continue;
+        }
+        printf("cost for box_id = %lu\n", box_id);
+        double sum = 0.0;
+        for (unsigned iy = 0; iy < matrix.dim_size(1); iy++) {
+            for (unsigned ix = 0; ix < matrix.dim_size(0); ix++) {
+                const auto& entry = matrix[ix][iy];
+                if (entry.valid()) {
+                    sum += entry.delay;
+                }
             }
-
-            int node_cost_ind = device_ctx.rr_nodes[node_ind].cost_index();
-            int node_seg_ind = device_ctx.rr_indexed_data[node_cost_ind].seg_index;
-            if (node_seg_ind == seg_index_) {
-                index_ += 1;
-                return node_ind;
+        }
+        double avg = sum / ((double)matrix.dim_size(0) * (double)matrix.dim_size(1));
+        for (unsigned iy = 0; iy < matrix.dim_size(1); iy++) {
+            for (unsigned ix = 0; ix < matrix.dim_size(0); ix++) {
+                const auto& entry = matrix[ix][iy];
+                if (entry.valid() && entry.delay > avg) {
+                    printf("o");
+                } else {
+                    printf(".");
+                }
             }
+            printf("\n");
         }
-
-        return UNDEFINED;
     }
+}
 
-  private:
-    int start_x_;
-    int start_y_;
-    t_rr_type rr_type_;
-    int seg_index_;
-    size_t index_;
-};
-
-// Minimum size of search for channels to profile.  kMinProfile results
-// in searching x = [0, kMinProfile], and y = [0, kMinProfile[.
-//
-// Making this value larger will increase the sample size, but also the runtime
-// to produce the lookahead.
-static constexpr int kMinProfile = 1;
-
-// Maximum size of search for channels to profile.  Once search is outside of
-// kMinProfile distance, lookahead will stop searching once:
-//  - At least one channel has been profiled
-//  - kMaxProfile is exceeded.
-static constexpr int kMaxProfile = 7;
-
-static int search_at(int iseg, int start_x, int start_y, t_routing_cost_map* cost_map) {
-    const auto& device_ctx = g_vpr_ctx.device();
+std::list<std::pair<int, int> > CostMap::list_empty() const {
```

Comment:
There is almost never a reason to use a linked list, except for very specific circumstances.  Why are you using a linked list here?

Reply:
It seemed the natural datatype, I haven't tried to optimize much yet.